modules/auxiliary/gather/ldap_query.rb

```diff
@@ -206,15 +301,113 @@ def output_data_csv(entries)
     generate_rex_tables(entries, 'csv')
   end
 
-  def normalize_entries(entries)
+  def query_attributes_data(ldap, entry_keys)
+    # def perform_ldap_query(ldap, filter, attributes, base: nil)
+    filter = '(|'
+    entry_keys.each do |key|
+      filter += "(LDAPDisplayName=#{key})"
+    end
+    filter += ')'
+    attributes = ['LDAPDisplayName', 'isSingleValued', 'oMSyntax', 'attributeSyntax']
+    attributes_data = perform_ldap_query(ldap, filter, attributes, base: ['CN=Schema,CN=Configuration', @base_dn].join(','))
+
+    entry_list = []
+    for entry in attributes_data do
+      data = {}
+      data[:ldapdisplayname] = entry[:ldapdisplayname][0]
+      if entry[:issinglevalued][0] == 'TRUE'
+        data[:issinglevalued] = 1
+      else
+        data[:issinglevalued] = 0
+      end
+      data[:omsyntax] = entry[:omsyntax][0].to_i
+      data[:attributesyntax] = entry[:attributesyntax][0]
+      entry_list << data
+    end
+    entry_list
+  end
+
+  def normalize_entries(ldap, entries)
     cleaned_entries = []
     entries.each do |entry|
       # Convert to a hash so we get only the data we need.
       entry = entry.to_h
       entry_keys = entry.keys
-      entry_keys.each do |key|
-        entry[key] = entry[key].map { |v| Rex::Text.to_hex_ascii(v) }
+      entry_keys.sort! # Sort entry keys alphabetically since LDAP results are returned alphabetically.
+      entry_list = query_attributes_data(ldap, entry_keys)
```

Comment:
You're querying the attribute information for each entry you receive. Each entry's attributes should be a subset of the attributes that were specified in the query. You should be able to do this once per query and not once per entry.

Reply:
Thanks for pointing this out, had to rework some stuff to take advantage of the hashes to deduplicate data but once I did think this is a nicer solution. Fix is in https://github.com/rapid7/metasploit-framework/pull/16979/commits/388934565494425e31efc60636e6dfd33f7ec664, and should make sure we only query for this information once. Hashes are used to get a simple set of all the attributes we need to retrieve for all the data entries returned with no overlap.